### ISSUE TITLE
slim.learning.train can't restore variables if new variable have created.

### ISSUE THREAD

**Issue by gauss-clb** [tensorflow/tensorflow#13342]
Look at my code:

```python
def slim_train_init_fn_test():
    '''
        In the model_ckpt/slim_train_init_fn_test.ckpt,
        {'x/y': 1.0, 'x/z': 2.0, 'y/z': 3.0}
    '''
    x_y = slim.variable('x/y', initializer=4.0)
    x_z = slim.variable('x/z', initializer=5.0)
    y_z = slim.variable('y/z', initializer=6.0)

    variables_to_restore = slim.get_variables_to_restore(include=['x'])

    init_fn = tf.contrib.framework.assign_from_checkpoint_fn(
        'model_ckpt/slim_train_init_fn_test.ckpt', variables_to_restore)
    optimizer = tf.train.GradientDescentOptimizer(learning_rate=.001)
    loss = slim.variable('loss', initializer=10.0)
    train_op = slim.learning.create_train_op(loss, optimizer)
    slim.learning.train(train_op, 'log', init_fn=init_fn, number_of_steps=1)
```

 In the `model_ckpt/slim_train_init_fn_test.ckpt`,  there is a map  `{'x/y': 1.0, 'x/z': 2.0, 'y/z': 3.0}`, and I use 
`get_variables_to_restore(include=['x'])` to restore variable `x_y` and  `x_z` which are in  `model_ckpt/slim_train_init_fn_test.ckpt`, but I create a new variable `loss` which isn't in `model_ckpt/slim_train_init_fn_test.ckpt`, the code raise an error:  Can't find key `loss` in check point file. 
So I can't understand why check point file should have key `loss`,  does `init_fn` not pass `variables_to_restore` to `tf.train.Saver` ? Is it a bug?

View the [assign_from_checkpoint_fn](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/framework/python/ops/variables.py#L659), it indeed uses `assign_from_checkpoint_fn` to initialize `tf.train.Saver` .  And there is another parameter `saver`  in  [slim.learning.train](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/slim/python/slim/learning.py#L729), it is  just used to [save](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/slim/python/slim/learning.py#L774)  parameters of model, not `restore`. 

So what's  wrong?

**Comment by gauss-clb** [comment]
After I view the source, I find there maybe  faulty design. [slim.learning.train](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/slim/python/slim/learning.py#L717) use `Supervisor`, and the [construct function](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/supervisor.py#L200) of `Supervisor` use [_init_saver](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/supervisor.py#L438) to initialize saver, which means if global_variables() is not NULL, saver will be `tf.train.saver()`.
And then it will create [SessionManager](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/supervisor.py#L342) and invoke [managed_session](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/supervisor.py#L889) and [prepare_or_wait_for_session](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/supervisor.py#L704),
but if the `saver` is not NULL, the `SessionManager` [won't run `init_op` and `init_fn`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/session_manager.py#L274), that's why it raise an error.

It means the `saver` is always not NULL, and `init_op` and `init_fn` will never run.

**Comment by Amitayus** [comment]
I have the same problem with you. I solve it by doing following two steps.

### 1. pass the parameter `saver` to `slim.learning.train()`
```python
ckpt = tf.train.get_checkpoint_state(FLAGS.train_dir)
saver = tf.train.Saver(var_list=optimistic_restore_vars(ckpt.model_checkpoint_path) if ckpt else None)
```
where function [optimistic_restore_vars](https://github.com/tensorflow/tensorflow/issues/312) is defined as
```python
def optimistic_restore_vars(model_checkpoint_path):
    reader = tf.train.NewCheckpointReader(model_checkpoint_path)
    saved_shapes = reader.get_variable_to_shape_map()
    var_names = sorted([(var.name, var.name.split(':')[0]) for var in tf.global_variables()
                        if var.name.split(':')[0] in saved_shapes])
    restore_vars = []
    name2var = dict(zip(map(lambda x:x.name.split(':')[0], tf.global_variables()), tf.global_variables()))
    with tf.variable_scope('', reuse=True):
        for var_name, saved_var_name in var_names:
            curr_var = name2var[saved_var_name]
            var_shape = curr_var.get_shape().as_list()
            if var_shape == saved_shapes[saved_var_name]:
                restore_vars.append(curr_var)
    return restore_vars
```

### 2. pass the parameter `local_init_op` to `slim.learning.train()` to initialize the added new variables
```python  
local_init_op = tf.global_variables_initializer()
```

### In last, the code should look like this
```python
ckpt = tf.train.get_checkpoint_state(FLAGS.train_dir)
saver = tf.train.Saver(var_list=optimistic_restore_vars(ckpt.model_checkpoint_path) if ckpt else None)
local_init_op = tf.global_variables_initializer()

###########################
# Kicks off the training. #
###########################
learning.train(
    train_tensor,
    saver=saver,
    local_init_op=local_init_op,
    logdir=FLAGS.train_dir,
    master=FLAGS.master,
    is_chief=(FLAGS.task == 0),
    init_fn=_get_init_fn(),
    summary_op=summary_op,
    number_of_steps=FLAGS.max_number_of_steps,
    log_every_n_steps=FLAGS.log_every_n_steps,
    save_summaries_secs=FLAGS.save_summaries_secs,
    save_interval_secs=FLAGS.save_interval_secs,
    sync_optimizer=optimizer if FLAGS.sync_replicas else None)
```

**Comment by gauss-clb** [comment]
Don't put checkpoint file in logdir, so that the saver can't restore and  `init_fn` will work. So just put checkpoint file in another directory and pass to `tf.contrib.framework.assign_from_checkpoint_fn`.  I  read the source and find the solution.

**Comment by Amitayus** [comment]
Yeah, slim will not invoke init_op, but I found the parameter  `local_init_op` is done before `saver.restore`. Thus, I use `local_init_op=tf.global_variables_initializer()` and pass it into `slim.learning.train()`. Is this a proper way? Thank you.

**Comment by Amitayus** [comment]
Ref in `tensorflow/python/training/Supervisor (about line 186):
  ##### Custom model initialization

  `managed_session()` only supports initializing the model by running an
  `init_op` or restoring from the latest checkpoint.  If you have special
  initialization needs, see how to specify a `local_init_op` when creating the
  supervisor.  You can also use the `SessionManager` directly to create a
  session and check if it could be initialized automatically.
  """

**Comment by yekeren** [comment]
I have the same problem, and solve it using session wrapper function.

```Python
  def _session_wrapper_fn(sess):
    feed_init_fn(sess)
    return sess

  slim.learning.train(train_op, 
      logdir=FLAGS.train_log_dir,
      graph=g,
      master='',
      is_chief=True,
      number_of_steps=train_config.number_of_steps,
      log_every_n_steps=train_config.log_every_n_steps,
      save_interval_secs=train_config.save_interval_secs,
      save_summaries_secs=train_config.save_summaries_secs,
      session_config=session_config,
      session_wrapper=_session_wrapper_fn,
      init_fn=init_fn,
      saver=saver)

```

**Comment by ChanZou** [comment]
@yekeren @gauss-clb 
I met the same problem but mine crashes when [a default saver is created](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/slim/python/slim/learning.py#L658)
 inside slim.learning.train for it requires all variables in the graph to be found in the checkpoint. If an explicit saver is defined using var_list, newly inserted variables won't be tracked and saved later on. May I know how did you work this out? Thanks!

**Comment by yekeren** [comment]
@ChanZou I noticed my case is slightly different. The data I fed in feed_init_fn() is the training data thus I do not have the need to save them later. For your case, I suggest you use the init_fn parameter of slim.learning.train. That is, as @gauss-clb said, ensure that there are no checkpoint files in logdir, then set init_fn to be the returned value of tf.contrib.framework.assign_from_checkpoint_fn.

**Comment by ChanZou** [comment]
@yekeren Thank you for your reply! Yes I missed that important trick to fool the saver. Now it works. Thanks!

**Comment by u10116032** [comment]
@gauss-clb First of all, thanks for your sharing. According to your comment "Don't put checkpoint file in logdir, so that the saver can't restore and init_fn will work." So in my concept, slim will still run init_op and init_fn, which means all global variables are initialized then.
However, in my case, I found that the additional variables (not included in checkpoint) are still not initialized in NON CHIEF worker, but the chief worker can train well (I think it means that all variables are initialized successfully).
Any ideas? Thanks!

**Comment by fperezgamonal** [comment]
Hello,

Recently I faced the same problem and I solved with the following steps:

1. As @Amitayus , I defined a `tf.saver()` (I get the variables to restore by using `optimistic_restore_vars` as well). This is done after defining the training op
```
vars2restore = optimistic_restore_vars(ckpt.model_checkpoint_path)
saver = tf.train.Saver(max_to_keep=3, keep_checkpoint_every_n_hours=2,
                                       var_list=vars2restore if checkpoint_path else None)
```
2.  Define `init_fn` as:
```
init_fn = tf.contrib.framework.assign_from_checkpoint_fn(checkpoint_path, vars2restore)
```
3. Pass both arguments to `slim.learning.train` as:
```
final_loss = slim.learning.train(
                    train_op,
                    log_dir,
                    # session_config=tf.ConfigProto(allow_soft_placement=True),
                    global_step=checkpoint_global_step_tensor,
                    save_summaries_secs=save_summaries_secs,
                    number_of_steps=training_schedule['max_iters'],
                    # init_fn=InitAssignFn,
                    # train_step_fn=train_step_fn,
                    saver=saver,
                    # local_init_op=local_init_op,
                    init_fn=init_fn,
                )
```
So, in my case `local_init_op` did not work as it is *probably* re-initialized the restored variables.
I am able to resume training with approximately the same loss reported before stopping/pausing it.
Notice that I still need to check if the custom saver is actually used since `init_fn` ends up creating one with the `var_list` passed.

Anyway, hopes it helps!

Cheers,
Ferran.
